File: 100.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

def eql(p,q):
    if p.val==q.val:
        left=False
        right=False
        if p.left and q.left:
            left=eql(p.left, q.left)
            if not left:
                return False
        if p.right and q.right:
            right=eql(p.right,q.right)
            if not right:
                return False
        if (not p.left) and (not q.left):
            left=True
        if (not p.right) and (not q.right):
            right=True
        if left and right:
            return True
        else:
            return False
    else:
        return False

class Solution:
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        if (not p) and (not q):
            return True
        if (not p) and q:
            return False
        if p and (not q):
            return False
        return eql(p, q)

File: xiong/100.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if(p == NULL && q == NULL) return true;
        if(p == NULL) return false;
        if(q == NULL) return false;
        deque<TreeNode*> s1, s2;
        s1.push_front(p);
        s2.push_front(q);
        while((!s1.empty()) && (!s2.empty())){
            if(s1.empty() || s2.empty()) return false;
            TreeNode* v1 = s1.back();
            TreeNode* v2 = s2.back();
            if(v1->val != v2->val) return false;
            s1.pop_back();
            s2.pop_back();
            if(v1->left != NULL) {
                s1.push_front(v1->left);
                if(v2->left == NULL) return false;
                s2.push_front(v2->left);
            }else if(v2->left != NULL) return false;
            if(v1->right != NULL){
                s1.push_front(v1->right);
                if(v2->right == NULL) return false;
                s2.push_front(v2->right);
            }else if(v2->right != NULL) return false;
        }
        return true;
        
    }
};